File: node/tests/wire/vbp/test_types.js
'use strict';

const { test } = require('node:test');
const assert = require('node:assert');
const {
  T_BOOL, T_INT2, T_INT4, T_INT8, T_FLOAT4, T_FLOAT8,
  T_TEXT, T_VARCHAR, T_BYTEA, T_UUID, T_DATE, T_TIME, T_TIMESTAMP,
  T_TIMESTAMPTZ, T_INTERVAL, T_NUMERIC, T_MONEY,
  T_JSON, T_JSONB, T_ARRAY, T_INET, T_MACADDR, T_CIDR,
  T_VECTOR, T_TSVECTOR, T_DOCUMENT, T_TS_POINT, T_GEO_POINT,
} = require('../../../src/wire/vbp/opcodes');
const {
  encodeValue, decodeValue, isKnownType, encodeInputParam, decodeOutputColumn,
  VBPTypeError, ENCODERS, DECODERS,
} = require('../../../src/wire/vbp/types');

test('types: T_BOOL encode/decode', () => {
  assert.deepStrictEqual(encodeValue(T_BOOL, true), Buffer.from([1]));
  assert.deepStrictEqual(encodeValue(T_BOOL, false), Buffer.from([0]));
  assert.strictEqual(decodeValue(T_BOOL, Buffer.from([1])), true);
  assert.strictEqual(decodeValue(T_BOOL, Buffer.from([0])), false);
});

test('types: T_INT4 round-trip', () => {
  const buf = encodeValue(T_INT4, 42);
  assert.strictEqual(buf.length, 4);
  assert.strictEqual(buf.readInt32LE(0), 42);
  assert.strictEqual(decodeValue(T_INT4, buf), 42);
  assert.strictEqual(decodeValue(T_INT4, encodeValue(T_INT4, -1)), -1);
  assert.strictEqual(decodeValue(T_INT4, encodeValue(T_INT4, 0)), 0);
  assert.strictEqual(decodeValue(T_INT4, encodeValue(T_INT4, 0x7FFFFFFF)), 0x7FFFFFFF);
  assert.strictEqual(decodeValue(T_INT4, encodeValue(T_INT4, -0x80000000)), -0x80000000);
});

test('types: T_INT8 round-trip', () => {
  const buf = encodeValue(T_INT8, 9007199254740991n);
  assert.strictEqual(decodeValue(T_INT8, buf), 9007199254740991);
});

test('types: T_FLOAT4/FLOAT8 round-trip', () => {
  // 3.14 is not exactly representable in IEEE-754; use near-equal.
  const f4 = decodeValue(T_FLOAT4, encodeValue(T_FLOAT4, 3.14));
  assert.ok(Math.abs(f4 - 3.14) < 1e-6, `f4=${f4}`);
  assert.strictEqual(decodeValue(T_FLOAT8, encodeValue(T_FLOAT8, 3.14159265358979)), 3.14159265358979);
});

test('types: T_TEXT length-prefix round-trip', () => {
  const buf = encodeValue(T_TEXT, 'hello');
  // u32 length (4 bytes) + 5 bytes = 9 bytes
  assert.strictEqual(buf.length, 9);
  assert.strictEqual(buf.readUInt32LE(0), 5);
  assert.strictEqual(decodeValue(T_TEXT, buf), 'hello');
});

test('types: T_VARCHAR alias of T_TEXT', () => {
  const a = encodeValue(T_TEXT, 'abc');
  const b = encodeValue(T_VARCHAR, 'abc');
  assert.deepStrictEqual(a, b);
});

test('types: T_BYTEA round-trip', () => {
  const data = Buffer.from([0x00, 0xFF, 0x10]);
  const buf = encodeValue(T_BYTEA, data);
  assert.deepStrictEqual(decodeValue(T_BYTEA, buf), data);
});

test('types: T_UUID string + binary', () => {
  const s = '550e8400-e29b-41d4-a716-446655440000';
  const buf = encodeValue(T_UUID, s);
  assert.strictEqual(buf.length, 16);
  assert.strictEqual(decodeValue(T_UUID, buf), s);
  // From raw 16-byte Buffer
  const raw = Buffer.from('550e8400e29b41d4a716446655440000', 'hex');
  assert.deepStrictEqual(encodeValue(T_UUID, raw), raw);
});

test('types: T_DATE encode (days since epoch)', () => {
  const buf = encodeValue(T_DATE, 0);
  assert.strictEqual(buf.length, 4);
  assert.strictEqual(buf.readInt32LE(0), 0);
});

test('types: T_TIME encode (microseconds)', () => {
  const buf = encodeValue(T_TIME, 0);
  assert.strictEqual(buf.length, 8);
  assert.strictEqual(Number(buf.readBigInt64LE(0)), 0);
});

test('types: T_TIMESTAMP encode (microseconds since epoch)', () => {
  const buf = encodeValue(T_TIMESTAMP, 0);
  assert.strictEqual(buf.length, 8);
});

test('types: T_TIMESTAMPTZ matches T_TIMESTAMP shape', () => {
  assert.strictEqual(encodeValue(T_TIMESTAMPTZ, 1000).length, 8);
});

test('types: T_INTERVAL round-trip', () => {
  const buf = encodeValue(T_INTERVAL, 1000000);
  assert.strictEqual(buf.length, 16);
});

test('types: T_NUMERIC as ASCII', () => {
  const buf = encodeValue(T_NUMERIC, '123.45');
  assert.strictEqual(decodeValue(T_NUMERIC, buf), '123.45');
});

test('types: T_MONEY round-trip (cents)', () => {
  const buf = encodeValue(T_MONEY, '1.00');
  // 100 cents
  assert.strictEqual(Number(buf.readBigInt64LE(0)), 100);
  assert.strictEqual(decodeValue(T_MONEY, buf), 1.0);
});

test('types: T_JSON round-trip', () => {
  const obj = { hello: 'world', n: 42 };
  const buf = encodeValue(T_JSON, obj);
  assert.deepStrictEqual(decodeValue(T_JSON, buf), obj);
});

test('types: T_JSONB = T_JSON shape', () => {
  assert.deepStrictEqual(encodeValue(T_JSONB, [1, 2]), encodeValue(T_JSON, [1, 2]));
});

test('types: T_ARRAY of int4', () => {
  const buf = encodeValue(T_ARRAY, [1, 2, 3, 4]);
  // u32 count (4) + 4 * i32 (16) = 20
  assert.strictEqual(buf.length, 20);
  assert.strictEqual(buf.readUInt32LE(0), 4);
  assert.deepStrictEqual(decodeValue(T_ARRAY, buf), [1, 2, 3, 4]);
});

test('types: T_INET, T_MACADDR, T_CIDR length-prefixed ASCII', () => {
  const inet = encodeValue(T_INET, '10.0.0.0/8');
  assert.strictEqual(decodeValue(T_INET, inet), '10.0.0.0/8');
  const mac = encodeValue(T_MACADDR, 'aa:bb:cc:dd:ee:ff');
  assert.strictEqual(decodeValue(T_MACADDR, mac), 'aa:bb:cc:dd:ee:ff');
  const cidr = encodeValue(T_CIDR, '192.168.0.0/16');
  assert.strictEqual(decodeValue(T_CIDR, cidr), '192.168.0.0/16');
});

test('types: T_VECTOR dim + f32×N', () => {
  const buf = encodeValue(T_VECTOR, [1.0, 2.0, 3.0]);
  assert.strictEqual(buf.length, 4 + 3 * 4);
  assert.strictEqual(buf.readUInt32LE(0), 3);
  const v = decodeValue(T_VECTOR, buf);
  assert.strictEqual(v.length, 3);
  assert.ok(Math.abs(v[0] - 1.0) < 1e-6);
});

test('types: T_VECTOR from {dim, values}', () => {
  const buf = encodeValue(T_VECTOR, { dim: 2, values: [0.5, 1.5] });
  assert.strictEqual(buf.readUInt32LE(0), 2);
});

test('types: T_TSVECTOR round-trip', () => {
  const buf = encodeValue(T_TSVECTOR, 'cat:1 dog:2');
  assert.strictEqual(decodeValue(T_TSVECTOR, buf), 'cat:1 dog:2');
});

test('types: T_DOCUMENT round-trip (JSON)', () => {
  const buf = encodeValue(T_DOCUMENT, { id: 1, name: 'foo' });
  assert.deepStrictEqual(decodeValue(T_DOCUMENT, buf), { id: 1, name: 'foo' });
});

test('types: T_TS_POINT (i64 micros + f64 value)', () => {
  const buf = encodeValue(T_TS_POINT, [1000000, 3.14]);
  assert.strictEqual(buf.length, 16);
  assert.strictEqual(Number(buf.readBigInt64LE(0)), 1000000);
  const [ts, v] = decodeValue(T_TS_POINT, buf);
  assert.strictEqual(ts, 1000000);
  assert.ok(Math.abs(v - 3.14) < 1e-9);
});

test('types: T_TS_POINT from {ts, value} dict', () => {
  const buf = encodeValue(T_TS_POINT, { ts: 42, value: 1.5 });
  const [ts, v] = decodeValue(T_TS_POINT, buf);
  assert.strictEqual(ts, 42);
  assert.strictEqual(v, 1.5);
});

test('types: T_GEO_POINT (i32 lat_e7 + i32 lon_e7)', () => {
  const buf = encodeValue(T_GEO_POINT, { lat: 37.7749, lon: -122.4194 });
  assert.strictEqual(buf.length, 8);
  // round-trip via array
  const [lat, lon] = decodeValue(T_GEO_POINT, buf);
  assert.strictEqual(lat, 377749000);
});

test('types: isKnownType returns true for registered IDs', () => {
  assert.strictEqual(isKnownType(T_BOOL), true);
  assert.strictEqual(isKnownType(T_INT4), true);
  assert.strictEqual(isKnownType(0xDEAD), false);
});

test('types: unknown type_id throws', () => {
  assert.throws(() => encodeValue(0xDEAD, 'x'), VBPTypeError);
  assert.throws(() => decodeValue(0xDEAD, Buffer.alloc(0)), VBPTypeError);
});

test('types: T_INT4 out-of-range throws', () => {
  assert.throws(() => encodeValue(T_INT4, 1e20), VBPTypeError);
});

test('types: input envelope with NULL', () => {
  const buf = encodeInputParam(T_INT4, null);
  // u16 type_id (2) + u8 null_tag=0 (1) = 3 bytes
  assert.strictEqual(buf.length, 3);
  assert.strictEqual(buf.readUInt16LE(0), T_INT4);
  assert.strictEqual(buf.readUInt8(2), 0);
});

test('types: input envelope with value', () => {
  const buf = encodeInputParam(T_INT4, 42);
  // u16 type_id (2) + u8 null_tag=1 (1) + i32 (4) = 7 bytes
  assert.strictEqual(buf.length, 7);
  assert.strictEqual(buf.readUInt16LE(0), T_INT4);
  assert.strictEqual(buf.readUInt8(2), 1);
  assert.strictEqual(buf.readInt32LE(3), 42);
});

test('types: input envelope with text', () => {
  const buf = encodeInputParam(T_TEXT, 'hi');
  // u16 type_id (2) + u8 null_tag=1 (1) + u32 len=2 (4) + 'hi' (2) = 9
  assert.strictEqual(buf.length, 9);
});

test('types: output column envelope decode', () => {
  // decodeOutputColumn input is the column body AFTER the u16 type_id:
  // u8 bitmap_count, [bitmap], values.
  const inner = Buffer.concat([
    Buffer.from([0]),                // bitmap_count = 0
    Buffer.from([42, 0, 0, 0]),     // value = 42 (i32 LE)
  ]);
  const { bitmap, values } = decodeOutputColumn(T_INT4, inner);
  assert.strictEqual(bitmap.length, 0);
  assert.strictEqual(values.length, 4);
  assert.strictEqual(values.readInt32LE(0), 42);
});

test('types: encoder/decoder dispatch tables are consistent', () => {
  assert.strictEqual(Object.keys(ENCODERS).length, Object.keys(DECODERS).length);
});

File: python/tests/wire/vbp/test_types.py
"""Tests for the vbp.wire.types module — encode/decode round-trips for all 27 v1 type IDs."""
from __future__ import annotations

import io
import os
import struct
import sys
import unittest
import uuid
from datetime import date, datetime, time, timedelta, timezone

# Make sure we can import the vedadb package from the repo root.
HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, "..", "..", ".."))
if REPO_ROOT not in sys.path:
    sys.path.insert(0, REPO_ROOT)

from vedadb.wire.vbp.opcodes import (  # noqa: E402
    T_ARRAY,
    T_BOOL,
    T_BYTEA,
    T_CIDR,
    T_DATE,
    T_DOCUMENT,
    T_FLOAT4,
    T_FLOAT8,
    T_GEO_POINT,
    T_INET,
    T_INT2,
    T_INT4,
    T_INT8,
    T_INTERVAL,
    T_JSON,
    T_JSONB,
    T_MACADDR,
    T_MONEY,
    T_NUMERIC,
    T_TEXT,
    T_TIME,
    T_TIMESTAMP,
    T_TIMESTAMPTZ,
    T_TS_POINT,
    T_TSVECTOR,
    T_UUID,
    T_VARCHAR,
    T_VECTOR,
    TYPE_IDS,
)
from vedadb.wire.vbp.types import (  # noqa: E402
    decode_value,
    encode_value,
    is_known_type,
)


class TestRegistry(unittest.TestCase):
    """The v1 type registry is 27 IDs as per the spec."""

    def test_exactly_27_type_ids(self):
        self.assertEqual(len(TYPE_IDS), 27)

    def test_all_in_registry_round_trip(self):
        for tid in TYPE_IDS:
            self.assertTrue(is_known_type(tid), f"type_id {tid} not in registry")

    def test_unknown_type_rejected(self):
        self.assertFalse(is_known_type(0xDEAD))
        with self.assertRaises(Exception):
            encode_value(0xDEAD, "x")
        with self.assertRaises(Exception):
            decode_value(0xDEAD, b"\x00" * 4)


class TestIntegerRoundTrip(unittest.TestCase):
    def test_bool(self):
        for v in (True, False):
            raw = encode_value(T_BOOL, v)
            self.assertEqual(decode_value(T_BOOL, raw), v)

    def test_int2_range(self):
        for v in (-32768, -1, 0, 1, 32767):
            raw = encode_value(T_INT2, v)
            self.assertEqual(decode_value(T_INT2, raw), v)

    def test_int4_range(self):
        for v in (-2**31, -1, 0, 1, 2**31 - 1):
            raw = encode_value(T_INT4, v)
            self.assertEqual(decode_value(T_INT4, raw), v)

    def test_int8_range(self):
        for v in (-(2**63), -1, 0, 1, 2**63 - 1):
            raw = encode_value(T_INT8, v)
            self.assertEqual(decode_value(T_INT8, raw), v)

    def test_int4_out_of_range(self):
        with self.assertRaises(Exception):
            encode_value(T_INT4, 2**32)


class TestFloatRoundTrip(unittest.TestCase):
    def test_float4(self):
        # float32 has ~7 significant digits; use values within that range.
        for v in (0.0, 1.5, -3.14, 1e6, 1e-6):
            raw = encode_value(T_FLOAT4, v)
            decoded = decode_value(T_FLOAT4, raw)
            # Use relative tolerance for larger values.
            if abs(v) >= 1.0:
                self.assertAlmostEqual(decoded / v, 1.0, places=4)
            else:
                self.assertAlmostEqual(decoded, v, places=5)

    def test_float8_full_precision(self):
        for v in (0.0, 1.5, -3.14159, 3.14159):
            raw = encode_value(T_FLOAT8, v)
            self.assertEqual(decode_value(T_FLOAT8, raw), v)


class TestStringByteRoundTrip(unittest.TestCase):
    def test_text_ascii(self):
        for s in ("", "hello", "x" * 1024):
            raw = encode_value(T_TEXT, s)
            self.assertEqual(decode_value(T_TEXT, raw), s)

    def test_text_unicode(self):
        for s in ("héllo", "日本語", "🚀", "héllo 日本語 🚀"):
            raw = encode_value(T_TEXT, s)
            self.assertEqual(decode_value(T_TEXT, raw), s)

    def test_varchar(self):
        s = "varchar test"
        self.assertEqual(decode_value(T_VARCHAR, encode_value(T_VARCHAR, s)), s)

    def test_bytea_with_nul(self):
        b = b"hello\x00world"
        raw = encode_value(T_BYTEA, b)
        self.assertEqual(decode_value(T_BYTEA, raw), b)

    def test_bytea_round_trip(self):
        for b in (b"", b"\x00", b"\xDE\xAD\xBE\xEF\x00\xFF", b"x" * 1024):
            self.assertEqual(decode_value(T_BYTEA, encode_value(T_BYTEA, b)), b)

    def test_uuid(self):
        u = uuid.UUID("12345678-1234-5678-1234-567812345678")
        raw = encode_value(T_UUID, u)
        self.assertEqual(decode_value(T_UUID, raw), u)

    def test_uuid_from_string(self):
        s = "12345678-1234-5678-1234-567812345678"
        raw = encode_value(T_UUID, s)
        self.assertEqual(decode_value(T_UUID, raw), uuid.UUID(s))


class TestDateTimeRoundTrip(unittest.TestCase):
    def test_date(self):
        d = date(2026, 6, 15)
        self.assertEqual(decode_value(T_DATE, encode_value(T_DATE, d)), d)

    def test_date_epoch(self):
        d = date(1970, 1, 1)
        self.assertEqual(decode_value(T_DATE, encode_value(T_DATE, d)), d)

    def test_time(self):
        t = time(12, 30, 45, 123456)
        self.assertEqual(decode_value(T_TIME, encode_value(T_TIME, t)), t)

    def test_timestamp(self):
        dt = datetime(2026, 6, 15, 12, 30, 45, 123456)
        self.assertEqual(decode_value(T_TIMESTAMP, encode_value(T_TIMESTAMP, dt)), dt)

    def test_timestamptz_normalizes_to_utc(self):
        # 2026-06-15T12:00:00+05:30 == 2026-06-15T06:30:00Z
        dt = datetime(2026, 6, 15, 12, 0, 0, tzinfo=timezone(timedelta(hours=5, minutes=30)))
        out = decode_value(T_TIMESTAMPTZ, encode_value(T_TIMESTAMPTZ, dt))
        self.assertEqual(out, datetime(2026, 6, 15, 6, 30, 0))

    def test_interval(self):
        td = timedelta(hours=2, minutes=30)
        out = decode_value(T_INTERVAL, encode_value(T_INTERVAL, td))
        self.assertEqual(out, td)


class TestNumericRoundTrip(unittest.TestCase):
    def test_numeric(self):
        for v in ("0", "123", "3.14159", "-1.5"):
            self.assertEqual(decode_value(T_NUMERIC, encode_value(T_NUMERIC, v)), v)

    def test_money(self):
        # T_MONEY is an optional v1 extension; the v1 closed set is 27
        # types and excludes T_MONEY.  We still ship a working encoder
        # for it but do not include it in TYPE_IDS.  The dispatch
        # table above therefore rejects T_MONEY (no entry); the
        # encode_value call below should raise.
        for v in (0.0, 1.23, -99.99):
            with self.assertRaises(Exception):
                encode_value(T_MONEY, v)


class TestStructuredRoundTrip(unittest.TestCase):
    def test_json(self):
        for v in (None, True, 42, "hello", [1, 2, 3], {"k": "v"}):
            self.assertEqual(decode_value(T_JSON, encode_value(T_JSON, v)), v)

    def test_jsonb(self):
        for v in (None, [1, 2, 3], {"a": 1, "b": [2, 3]}):
            self.assertEqual(decode_value(T_JSONB, encode_value(T_JSONB, v)), v)

    def test_array(self):
        v = [1, 2, 3, 4, 5]
        self.assertEqual(decode_value(T_ARRAY, encode_value(T_ARRAY, v)), v)

    def test_inet(self):
        self.assertEqual(decode_value(T_INET, encode_value(T_INET, "192.168.1.1")), "192.168.1.1")

    def test_macaddr(self):
        self.assertEqual(decode_value(T_MACADDR, encode_value(T_MACADDR, "00:1A:2B:3C:4D:5E")), "00:1A:2B:3C:4D:5E")

    def test_cidr(self):
        self.assertEqual(decode_value(T_CIDR, encode_value(T_CIDR, "10.0.0.0/8")), "10.0.0.0/8")


class TestVectorAndDocument(unittest.TestCase):
    def test_vector_from_list(self):
        v = [0.1, 0.2, 0.3]
        out = decode_value(T_VECTOR, encode_value(T_VECTOR, v))
        for a, b in zip(v, out):
            self.assertAlmostEqual(a, b, places=4)

    def test_vector_from_dict(self):
        v = {"dim": 4, "values": [1.0, 2.0, 3.0, 4.0]}
        out = decode_value(T_VECTOR, encode_value(T_VECTOR, v))
        self.assertEqual(len(out), 4)
        self.assertAlmostEqual(out[0], 1.0)
        self.assertAlmostEqual(out[3], 4.0)

    def test_document(self):
        v = {"fields": [{"name": "x", "value": 1}]}
        self.assertEqual(decode_value(T_DOCUMENT, encode_value(T_DOCUMENT, v)), v)

    def test_ts_point(self):
        v = (1718455845000000, 42.5)
        out = decode_value(T_TS_POINT, encode_value(T_TS_POINT, v))
        self.assertEqual(out, v)

    def test_ts_point_from_dict(self):
        v = {"ts": 1718455845000000, "value": 42.5}
        out = decode_value(T_TS_POINT, encode_value(T_TS_POINT, v))
        self.assertEqual(out, (1718455845000000, 42.5))

    def test_geo_point(self):
        v = (37747900, -122419400)
        out = decode_value(T_GEO_POINT, encode_value(T_GEO_POINT, v))
        self.assertEqual(out, v)

    def test_geo_point_from_dict(self):
        v = {"lat_e7": 37747900, "lon_e7": -122419400}
        out = decode_value(T_GEO_POINT, encode_value(T_GEO_POINT, v))
        self.assertEqual(out, (37747900, -122419400))


if __name__ == "__main__":
    unittest.main()
